### PR TITLE
Portenta camera fixes

### PR DESCRIPTION
**Pull request by iabdalkader**
* Fix himax default registers.
* Fix HIMAX_Open/boot.
* Fix broken DMA transfers in camera.cpp.
* Fix timer frequency (was running at 3MHz).
* Fix cache issues.
* Fix broken snapshot/grab.
* Fix wrong ConfigCROP args.
* Update API as requested.
* Update Arduino example sketch.
* Update Processing example.
* Add support for motion detection.
* Add motion detection example.

~~Note:
When the sensor enters standby mode after a frame grab, the auto-exposure function does Not update or gets reset, for now I've added that skip_frames() function to run enough frames for auto-exposure to kick in... but I suggest running the sensor in "normal" mode and adding a function to enter low-power standby() when needed.~~

### CODE REVIEW

**Comment by hpssjellis** [comment]
@iabdalkader: To test this do I just need to replace the ```Himax_HM01B0```  and  ```Portenta_Camera``` libraries with your versions [here](https://github.com/iabdalkader/ArduinoCore-mbed/tree/camera_fixes/libraries) ?

I am fine with the sketches and using MIT Processes.

**Comment by iabdalkader** [comment]
> @iabdalkader: To test this do I just need to replace the `Himax_HM01B0` and `Portenta_Camera` libraries with your versions [here](https://github.com/iabdalkader/ArduinoCore-mbed/tree/camera_fixes/libraries) ?
> 
> I am fine with the sketches and using MIT Processes.

Hi, yes, you need to replace camera.cpp, camera.h, himax.cpp and both the Arduino sketch and the Processing sketch. If the high baudrate doesn't work for you try 115200...Note I reverted some of the older registers and force pushed to the branch, so refresh to get the latest update.

**Comment by hpssjellis** [comment]
@iabdalkader 

Looks great using ubuntu to run processes even with the fast serial speed

c++camera

.

.

.

Absolute mess, as always, running processes on windows 10
c++camera-windows


Thanks for your work, I can take this now and run some machine learning programs using EdgeImpulse

P.S. Any chance of some lower resolution, (QQQVGA) or cropping functions in the library?

**Comment by iabdalkader** [comment]
> Looks great using ubuntu to run processes even with the fast serial speed

Great, just keep in mind the API may change I don't know if this PR will get merged as it is.

> Absolute mess, as always, running processes on windows 10

Looks like it might be reading less than requested, try looping until the whole frame is read or maybe it's overflowing...There were settings for this in Windows, for COM ports FIFO and buffer sizes try increasing it.
Maybe this I'm not sure...
image

> P.S. Any chance of some lower resolution, (QQQVGA) or cropping functions in the library?

Yes more fixes will likely happen in following PRs.

**Comment by hpssjellis** [comment]
Thanks @iabdalkader I ran an impulse microcontroller detection program and got these results
is-microcontroller

93% microcontroller using a 48x48 crop. 

Your new code has made that happen. Thanks so much.

File: libraries/Portenta_Camera/camera.cpp [hunk 1]
```diff
@@ -536,59 +533,65 @@ int CameraClass::begin(int horizontalResolution, int verticalResolution)
...
-  /* Wait until camera frame is ready : DCMI Frame event */
-  while(camera_frame_ready == 0)
-  {
+  BSP_CAMERA_SnapshotStart(buffer);
+
+  while (n_frames--) {
+      camera_frame_ready = 0;
+      /* Wait until camera frame is ready : DCMI Frame event */
+      for (uint32_t start = millis(); camera_frame_ready == 0;) {
+          __WFI();
```
[the remark this answers is not in the document]
Comment (iabdalkader) [reply libraries/Portenta_Camera/camera.cpp:581]: But it is a wait for interrupt, it should wake up on the first interrupt. Do you have something else in mind ?

File: libraries/Portenta_Camera/examples/Envie_camera/Envie_camera.ino [hunk 2]
```diff
@@ -1,21 +1,26 @@
...
+
+  // Init the cam
+  cam.begin(320, 240);
 
-  // put your setup code here, to run once:
-  cam.begin(324, 244);
-  cam.start();
-  //cam.testPattern(true);
+  // Skip 60 frames
+  cam.skip_frames(fb, 60);
```
[the remark this answers is not in the document]
Comment (iabdalkader) [reply libraries/Portenta_Camera/examples/Envie_camera/Envie_camera.ino:15]: Yes exactly, it seems when the sensor is put in standby mode at the end of grab() it messes up the auto-exposure, this is just to allow enough frames to run uninterrupted. I would suggest just leaving the sensor in streaming mode and adding a standby(bool enable) function to the API ?

File: libraries/Portenta_Camera/camera.cpp [hunk 3]
```diff
@@ -536,59 +533,65 @@ int CameraClass::begin(int horizontalResolution, int verticalResolution)
...
 
   /* Stop the camera to avoid having the DMA2D work in parallel of Display */
   /* which cause perturbation of LTDC                                      */
   BSP_CAMERA_Suspend();
-
-  return (uint8_t *)LCD_FRAME_BUFFER;
+  return 0;
 }
 
 void CameraClass::testPattern(bool walking)
```
[the remark this answers is not in the document]
Comment (iabdalkader) [reply libraries/Portenta_Camera/camera.cpp:597]: Will copy over the final registers from openmv...

File: libraries/Portenta_Camera/camera.h [hunk 4]
```diff
@@ -1,8 +1,7 @@
 class CameraClass {
 public:
 	int begin(int horizontalResolution, int verticalResolution);
-	uint8_t* snapshot();
-	int start(uint32_t timeout);
-	uint8_t* grab(void);
+	int grab(uint8_t *buffer, uint32_t timeout=5000);
```
[the remark this answers is not in the document]
Comment (iabdalkader) [reply libraries/Portenta_Camera/camera.h]: Added set_framerate() and standby(), manual exposure still doesn't work.

File: libraries/Portenta_Camera/camera.h [hunk 5]
```diff
@@ -1,8 +1,7 @@
...
 	void testPattern(bool walking);
-};
\ No newline at end of file
+	int skip_frames(uint8_t *buffer, uint32_t n_frames, uint32_t timeout=5000);
```
[the remark this answers is not in the document]
Comment (iabdalkader) [reply libraries/Portenta_Camera/camera.h]: Removed.

**Comment by hpssjellis** [comment]
@iabdalkader @pnndra can you guys check to see if the Portenta Camera works using the M4 Core as well as the M7 Core. These feature would be very useful. My preliminary test says it doesn't but I am checking some more.

**Comment by iabdalkader** [comment]
This driver isn't meant to be portable, it can be though.

**Comment by facchinm** [comment]
Thanks @iabdalkader , I merged it (was working perfectly), just changed a couple of APIs for motion detection to follow Arduino style guidelines.
This is a broader topic that should be properly tackled in ArduinoCore-API repo but I think we can live with these solution for the time being.

**Pull request merged**